File: frontend/packages/arch/api-schema/src/idl/rerank/space_rerank.ts
/*
 * Copyright 2025 ynet-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createAPI } from './../../api/config';

/** Rerank类型 */
export enum RerankType {
  OpenAI = 'openai',
}

/** Rerank状态 */
export enum RerankStatus {
  Enabled = 1,
  Disabled = 2,
}

/** OpenAI Rerank配置 */
export interface OpenAIRerankConfig {
  base_url: string;
  api_key: string;
  model: string;
}

/** 统一Rerank配置 */
export interface RerankConfig {
  type: RerankType;
  openai_config?: OpenAIRerankConfig;
}

/** 空间Rerank配置 */
export interface SpaceRerankConfig {
  id: string;
  space_id: string;
  name: string;
  description?: string;
  rerank_type: RerankType;
  config?: RerankConfig;
  status: RerankStatus;
  is_default: boolean;
  created_at: number;
  updated_at: number;
}

/** 创建空间Rerank配置请求 */
export interface CreateSpaceRerankRequest {
  space_id: string;
  name: string;
  description?: string;
  type: RerankType;
  config: Record<string, any>;
  set_as_default?: boolean;
}

export interface CreateSpaceRerankResponse {
  code: number;
  msg: string;
  data?: SpaceRerankConfig;
}

/** 获取空间Rerank列表请求 */
export interface ListSpaceReranksRequest {
  space_id: string;
}

export interface ListSpaceReranksResponse {
  code: number;
  msg: string;
  data?: SpaceRerankConfig[];
}

/** 获取空间默认Rerank请求 */
export interface GetSpaceDefaultRerankRequest {
  space_id: string;
}

export interface GetSpaceDefaultRerankResponse {
  code: number;
  msg: string;
  data?: SpaceRerankConfig;
}

/** 更新空间Rerank配置请求 */
export interface UpdateSpaceRerankRequest {
  space_id: string;
  rerank_id: string;
  name?: string;
  description?: string;
  type?: RerankType;
  config?: Record<string, any>;
}

export interface UpdateSpaceRerankResponse {
  code: number;
  msg: string;
  data?: SpaceRerankConfig;
}

/** 删除空间Rerank配置请求 */
export interface DeleteSpaceRerankRequest {
  space_id: string;
  rerank_id: string;
}

export interface DeleteSpaceRerankResponse {
  code: number;
  msg: string;
}

/** 设置默认Rerank请求 */
export interface SetDefaultSpaceRerankRequest {
  space_id: string;
  rerank_id: string;
}

export interface SetDefaultSpaceRerankResponse {
  code: number;
  msg: string;
}

/** 启用/禁用Rerank请求 */
export interface EnableSpaceRerankRequest {
  space_id: string;
  rerank_id: string;
}

export interface EnableSpaceRerankResponse {
  code: number;
  msg: string;
}

export interface DisableSpaceRerankRequest {
  space_id: string;
  rerank_id: string;
}

export interface DisableSpaceRerankResponse {
  code: number;
  msg: string;
}

// API 函数定义

export const CreateSpaceRerank = /*#__PURE__*/createAPI<CreateSpaceRerankRequest, CreateSpaceRerankResponse>({
  url: '/api/rerank/space/create',
  method: 'POST',
  name: 'CreateSpaceRerank',
  reqType: 'CreateSpaceRerankRequest',
  reqMapping: {
    body: ['space_id', 'name', 'description', 'type', 'config', 'set_as_default'],
  },
  resType: 'CreateSpaceRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const ListSpaceReranks = /*#__PURE__*/createAPI<ListSpaceReranksRequest, ListSpaceReranksResponse>({
  url: '/api/rerank/space/list',
  method: 'POST',
  name: 'ListSpaceReranks',
  reqType: 'ListSpaceReranksRequest',
  reqMapping: {
    body: ['space_id'],
  },
  resType: 'ListSpaceReranksResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const GetSpaceDefaultRerank = /*#__PURE__*/createAPI<GetSpaceDefaultRerankRequest, GetSpaceDefaultRerankResponse>({
  url: '/api/rerank/space/default',
  method: 'POST',
  name: 'GetSpaceDefaultRerank',
  reqType: 'GetSpaceDefaultRerankRequest',
  reqMapping: {
    body: ['space_id'],
  },
  resType: 'GetSpaceDefaultRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const UpdateSpaceRerank = /*#__PURE__*/createAPI<UpdateSpaceRerankRequest, UpdateSpaceRerankResponse>({
  url: '/api/rerank/space/update',
  method: 'POST',
  name: 'UpdateSpaceRerank',
  reqType: 'UpdateSpaceRerankRequest',
  reqMapping: {
    body: ['space_id', 'rerank_id', 'name', 'description', 'type', 'config'],
  },
  resType: 'UpdateSpaceRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const DeleteSpaceRerank = /*#__PURE__*/createAPI<DeleteSpaceRerankRequest, DeleteSpaceRerankResponse>({
  url: '/api/rerank/space/delete',
  method: 'POST',
  name: 'DeleteSpaceRerank',
  reqType: 'DeleteSpaceRerankRequest',
  reqMapping: {
    body: ['space_id', 'rerank_id'],
  },
  resType: 'DeleteSpaceRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const SetDefaultSpaceRerank = /*#__PURE__*/createAPI<SetDefaultSpaceRerankRequest, SetDefaultSpaceRerankResponse>({
  url: '/api/rerank/space/set-default',
  method: 'POST',
  name: 'SetDefaultSpaceRerank',
  reqType: 'SetDefaultSpaceRerankRequest',
  reqMapping: {
    body: ['space_id', 'rerank_id'],
  },
  resType: 'SetDefaultSpaceRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const EnableSpaceRerank = /*#__PURE__*/createAPI<EnableSpaceRerankRequest, EnableSpaceRerankResponse>({
  url: '/api/rerank/space/enable',
  method: 'POST',
  name: 'EnableSpaceRerank',
  reqType: 'EnableSpaceRerankRequest',
  reqMapping: {
    body: ['space_id', 'rerank_id'],
  },
  resType: 'EnableSpaceRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

export const DisableSpaceRerank = /*#__PURE__*/createAPI<DisableSpaceRerankRequest, DisableSpaceRerankResponse>({
  url: '/api/rerank/space/disable',
  method: 'POST',
  name: 'DisableSpaceRerank',
  reqType: 'DisableSpaceRerankRequest',
  reqMapping: {
    body: ['space_id', 'rerank_id'],
  },
  resType: 'DisableSpaceRerankResponse',
  schemaRoot: 'api://schemas/idl_rerank_space_rerank',
  service: 'space_rerank',
});

File: backend/domain/rerank/entity/space_rerank.go
/*
 * Copyright 2025 ynet-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"encoding/json"
	"strconv"
)

// RerankType represents the type of rerank provider
type RerankType string

const (
	RerankTypeOpenAI RerankType = "openai"
)

// SpaceRerank represents the space rerank configuration entity
type SpaceRerank struct {
	ID         uint64     `gorm:"primaryKey;autoIncrement;comment:主键ID" json:"id"`
	SpaceID    uint64     `gorm:"not null;comment:空间ID" json:"space_id"`
	UserID     uint64     `gorm:"not null;comment:创建者ID" json:"user_id"`
	Name       string     `gorm:"size:128;not null;comment:配置名称" json:"name"`
	Description string    `gorm:"size:512;default:'';comment:配置描述" json:"description"`
	RerankType RerankType `gorm:"size:32;not null;comment:Rerank类型" json:"rerank_type"`
	Config     string     `gorm:"type:json;not null;comment:Rerank配置" json:"config"`
	Status     int        `gorm:"not null;default:1;comment:状态: 1启用 2禁用" json:"status"`
	IsDefault  int        `gorm:"not null;default:0;comment:是否为默认配置" json:"is_default"`
	CreatedAt  uint64     `gorm:"not null;default:0;comment:创建时间" json:"created_at"`
	UpdatedAt  uint64     `gorm:"not null;default:0;comment:更新时间" json:"updated_at"`
	DeletedAt  *uint64    `gorm:"comment:删除时间" json:"deleted_at"`
}

func (SpaceRerank) TableName() string {
	return "space_rerank"
}

// Status constants
const (
	SpaceRerankStatusEnabled  = 1
	SpaceRerankStatusDisabled = 2
)

// OpenAIRerankConfig represents OpenAI-compatible rerank configuration
type OpenAIRerankConfig struct {
	BaseURL string `json:"base_url"`
	APIKey  string `json:"api_key"`
	Model   string `json:"model"`
}

// RerankConfig represents the unified rerank configuration
type RerankConfig struct {
	Type   RerankType          `json:"type"`
	OpenAI *OpenAIRerankConfig `json:"openai_config,omitempty"`
}

// GetConfigStruct parses the JSON config string into RerankConfig
func (s *SpaceRerank) GetConfigStruct() (*RerankConfig, error) {
	var config RerankConfig
	if err := json.Unmarshal([]byte(s.Config), &config); err != nil {
		return nil, err
	}
	config.Type = s.RerankType
	return &config, nil
}

// SetConfigFromStruct converts RerankConfig to JSON string and sets it
func (s *SpaceRerank) SetConfigFromStruct(config *RerankConfig) error {
	data, err := json.Marshal(config)
	if err != nil {
		return err
	}
	s.Config = string(data)
	s.RerankType = config.Type
	return nil
}

// SpaceRerankView is a view model for API responses
type SpaceRerankView struct {
	ID         string        `json:"id"`
	SpaceID    string        `json:"space_id"`
	Name       string        `json:"name"`
	Description string       `json:"description"`
	RerankType RerankType    `json:"rerank_type"`
	Config     *RerankConfig `json:"config"`
	Status     int           `json:"status"`
	IsDefault  bool          `json:"is_default"`
	CreatedAt  uint64        `json:"created_at"`
	UpdatedAt  uint64        `json:"updated_at"`
}

// ToView converts SpaceRerank to SpaceRerankView
func (s *SpaceRerank) ToView() *SpaceRerankView {
	config, _ := s.GetConfigStruct()

	// Mask sensitive information
	if config != nil {
		if config.OpenAI != nil && config.OpenAI.APIKey != "" {
			config.OpenAI.APIKey = maskAPIKey(config.OpenAI.APIKey)
		}
	}

	return &SpaceRerankView{
		ID:          formatID(s.ID),
		SpaceID:     formatID(s.SpaceID),
		Name:        s.Name,
		Description: s.Description,
		RerankType:  s.RerankType,
		Config:      config,
		Status:      s.Status,
		IsDefault:   s.IsDefault == 1,
		CreatedAt:   s.CreatedAt,
		UpdatedAt:   s.UpdatedAt,
	}
}

// maskAPIKey masks the API key for security
func maskAPIKey(key string) string {
	if len(key) <= 8 {
		return "****"
	}
	return key[:4] + "****" + key[len(key)-4:]
}

// formatID converts uint64 to string
func formatID(id uint64) string {
	return strconv.FormatUint(id, 10)
}
